itwin-platform-access/imodels-access-backend/src/interface-adapters/ErrorAdapter.ts

```diff
@@ -0,0 +1,139 @@
+    if (!errorNumber)
+      errorNumber = ErrorAdapter.mapErrorCode(error.code);
+
+    return new IModelError(errorNumber, error.message);
+  }
+
+  private static isAPIAuthError(apiErrorCode: IModelsErrorCode): boolean {
+    switch (apiErrorCode) {
+      case IModelsErrorCode.Unauthorized:
+      case IModelsErrorCode.InsufficientPermissions:
+        return true;
+      default:
+        return false;
+    }
+  }
+
+  private static isIncorrectAPIUsageError(apiErrorCode: IModelsErrorCode): boolean {
+    switch (apiErrorCode) {
+      case IModelsErrorCode.TooManyRequests:
+      case IModelsErrorCode.RequestTooLarge:
+      case IModelsErrorCode.InvalidIModelsRequest:
+      case IModelsErrorCode.InvalidValue:
+      case IModelsErrorCode.InvalidHeaderValue:
+      case IModelsErrorCode.InvalidRequestBody:
+      case IModelsErrorCode.InvalidThumbnailFormat:
+      case IModelsErrorCode.MutuallyExclusivePropertiesProvided:
+      case IModelsErrorCode.MissingRequestBody:
+      case IModelsErrorCode.MissingRequiredProperty:
+      case IModelsErrorCode.MissingRequiredParameter:
+      case IModelsErrorCode.MissingRequiredHeader:
+      case IModelsErrorCode.InvalidChange: // returned when, for example, user attempts to complete Baseline file upload while it is not in progress
+      case IModelsErrorCode.DataConflict: // returned when, for example, user attempts to complete Baseline file upload and the declared file size does not match actual uploaded file size
+        return true;
+      default:
+        return false;
+    }
+  }
+
+  private static isAPIErrorWithoutCorrespondingStatus(apiErrorCode: IModelsErrorCode): boolean {
+    switch (apiErrorCode) {
+      case IModelsErrorCode.NamedVersionNotFound:
+      case IModelsErrorCode.UserNotFound:
+      case IModelsErrorCode.BaselineFileNotFound:
+      case IModelsErrorCode.BaselineFileInitializationFailed:
+      case IModelsErrorCode.IModelFromTemplateInitializationFailed:
+      case IModelsErrorCode.ChangesetDownloadFailed:
+        return true;
+      default: return false;
+    }
+  }
+
+  private static tryMapGenericErrorCodeBasedOnOperation(
+    apiErrorCode: IModelsErrorCode,
+    operationName?: OperationNameForErrorMapping
+  ): IModelHubStatus | ChangeSetStatus | undefined {
+    if (!operationName)
+      return;
+
+    if (apiErrorCode === IModelsErrorCode.ResourceQuotaExceeded && operationName === "acquireBriefcase")
+      return IModelHubStatus.MaximumNumberOfBriefcasesPerUser;
+
+    if (apiErrorCode === IModelsErrorCode.RateLimitExceeded && operationName === "acquireBriefcase")
+      return IModelHubStatus.MaximumNumberOfBriefcasesPerUserPerMinute;
+
+    if (apiErrorCode === IModelsErrorCode.DownloadAborted && operationName === "downloadChangesets")
+      return ChangeSetStatus.DownloadCancelled;
+
+    return undefined;
+  }
+
+  private static mapErrorCode(apiErrorCode: IModelsErrorCode): IModelHubStatus {
```

Comment:
@swwilson-bsi please take a look and see whether these error mappings make sense, especially ones in `isAPIErrorWithoutCorrespondingStatus`. Nonexistent named version is handled already in `BackendIModelsAccess` [here](https://github.com/iTwin/imodels-clients/blob/cf0b35d8c024df8ca3562b8bebce5fb38fc56cc5/itwin-platform-access/imodels-access-backend/src/BackendIModelsAccess.ts#L171C26-L171C26), it returns `IModelStatus`, this does not seem entirely correct but this is probably because there is no appropriate status in `IModelHubStatus`.

also do you think we need similar error handing in `FrontendIModelsAccess` from @itwin/imodels-access-frontend?

regarding operation name in `tryMapGenericErrorCodeBasedOnOperation` - I have added this additional parameter because the returned error codes to me sound generic and like they could potentially be reused in other operations. I chose to handle error code together with operation name to avoid accidentally breaking the client in case we reuse this code. @Animagne do you think this is excessive?

Reply:
@calebmshafer @pmconne should review these mappings.

Reply:
This looks OK to me